dags/person_overrides.py

```diff
@@ -0,0 +1,482 @@
+
+    def create(self, client: Client, shards: int, max_execution_time: int) -> None:
+        client.execute(
+            f"""
+            CREATE DICTIONARY IF NOT EXISTS {self.qualified_name} (
+                team_id Int64,
+                distinct_id String,
+                person_id UUID,
+                version Int64
+            )
+            PRIMARY KEY team_id, distinct_id
+            SOURCE(CLICKHOUSE(DB %(database)s TABLE %(table)s PASSWORD %(password)s))
+            LAYOUT(COMPLEX_KEY_HASHED(SHARDS {shards}))
+            LIFETIME(0)
+            SETTINGS(max_execution_time={max_execution_time})
+            """,
+            {
+                "database": settings.CLICKHOUSE_DATABASE,
+                "table": self.source.name,
+                "password": settings.CLICKHOUSE_PASSWORD,
+            },
+        )
+
+    def exists(self, client: Client) -> bool:
+        results = client.execute(
+            "SELECT count() FROM system.dictionaries WHERE database = %(database)s AND name = %(name)s",
+            {"database": settings.CLICKHOUSE_DATABASE, "name": self.name},
+        )
+        [[count]] = results
+        return count > 0
+
+    def drop(self, client: Client) -> None:
+        client.execute(f"DROP DICTIONARY IF EXISTS {self.qualified_name} SYNC")
+
+    def __is_loaded(self, client: Client) -> bool:
+        results = client.execute(
+            "SELECT status, last_exception FROM system.dictionaries WHERE database = %(database)s AND name = %(name)s",
+            {"database": settings.CLICKHOUSE_DATABASE, "name": self.name},
+        )
+        if not results:
+            raise Exception("dictionary does not exist")
+        else:
+            [[status, last_exception]] = results
+            if status == "LOADED":
+                return True
+            elif status in {"LOADING", "FAILED_AND_RELOADING", "LOADED_AND_RELOADING"}:
+                return False
+            elif status == "FAILED":
+                raise Exception(f"failed to load: {last_exception}")
+            else:
+                raise Exception(f"unexpected status: {status}")
+
+    def load(self, client: Client):
+        # TODO: this should probably not reload if the dictionary is already loaded
+        client.execute(f"SYSTEM RELOAD DICTIONARY {self.qualified_name}")
+
+        # reload is async, so we need to wait for the dictionary to actually be loaded
+        # TODO: this should probably throw on unexpected reloads
+        while not self.__is_loaded(client):
+            time.sleep(5.0)
+
+        results = client.execute(
+            f"""
+            SELECT groupBitXor(row_checksum) AS table_checksum
+            FROM (SELECT cityHash64(*) AS row_checksum FROM {self.qualified_name} ORDER BY team_id, distinct_id)
+            """
+        )
+        [[checksum]] = results
+        return checksum
+
+    def enqueue_person_id_update_mutation(self, client: Client) -> Mutation:
```

Comment:
This could be abstracted along with the method below (there's a fair amount of duplication here), but not sure it's worth it.

Reply:
Cleaned this up a little bit with cce62ffe3bb5ee259023a2b07e6692f42c20e7d8 - thought about splitting out into something separate but since this is pretty dependent on the "find a mutation using this specifically named dictionary" trick to work correctly, I think it makes sense to keep it private-ish since generalizing it further would take a lot of work to make it reliably correct.